Torch baselines for attention benchmarks

Priya2698 opened the pull request:
Issue #1668

The benchmarks use `dropout_p=0.2` for benchmarking. For validating in forward pass, to ensure the same dropout_mask is applied, this is set to 0.

naoyam reviewed (approved):
LGTM

liqiangxl commented on benchmarks/python/test_huggingface_attn_bwd.py:
```diff
@@ -1,7 +1,7 @@
 from nvfuser import FusionDefinition, DataType
 from nvfuser.pytorch_utils import torch_dtype_to_nvfuser_dtype
-from .core import run_benchmark, clear_cuda_cache
+from .core import run_benchmark, clear_cuda_cache, unary_bwd_torch
```
where does this `unary_bwd_torch` defined? It doesn't exist in `core.py`.

Priya2698 replied:
It is defined here: https://github.com/NVIDIA/Fuser/blob/e10c4714aede47a41a1d71a879681d3f94c00df0/benchmarks/python/core.py#L142

liqiangxl commented on benchmarks/python/test_huggingface_attn_bwd.py:
```diff
@@ -95,3 +101,34 @@ def test_huggingface_attn_bwd_benchmark(
+        benchmark,
+        torch.compile(unary_bwd_torch) if compile else unary_bwd_torch,
+        [output, grads],
+        iobytes=huggingface_attn_bwd_iobytes(size, dtype),
```
can you comment why we need to set `iobytes`? what's the difference is using default.

Priya2698 replied:
This is for uniformity of IObytes between nvFuser and torch when the inputs/outputs differ between the two. See the discussion at: https://github.com/NVIDIA/Fuser/pull/1725. I'll add a reference to this PR in a comment.

liqiangxl commented on benchmarks/python/test_huggingface_attn_fwd.py:
```diff
@@ -71,9 +71,26 @@ def huggingface_attn_fwd_fusion(
     fd.add_output(T47)
 
 
+def huggingface_attn_fwd_fn(
```
do we have references for this and all the other baseline implementations?

Priya2698 replied:
I'll add links to the thunder code: They were removed earlier. 
Is that what you're referring to?

liqiangxl commented on benchmarks/python/test_huggingface_attn_fwd.py:
```diff
@@ -103,3 +120,28 @@ def test_huggingface_attn_fwd_benchmark(
+    clear_cuda_cache()
+
+    batch_size, seq_len, nh, n_embd = size
+    dropout_p = 0.0
```
is there an influence on perf if dropout is larger than 0, e.g. 0.2? If has, we can use a more practical value and set to 0 if validation is turned on.

liqiangxl replied:
If its `0`, the predication is trivial and may influence the perf.
```
   S9 = fd.define_scalar(1 - dropout_p, dtype=DataType.Double)
    T10 = fd.ops.lt(T8, S9)
```

Priya2698 replied:
Right. Let me make that change. Thanks!

Priya2698 replied:
This would although mean that we benchmark different things based on whether validation is turned on.

liqiangxl replied:
can you check the influence on bandwidth? I am not sure how much it may have, maybe very small.

Priya2698 replied:
What do you think about having another fusion definition with 0.0 dropout in validation and 0.2 otherwise. That would ensure we always benchmark with dropout (0.2) and it is consistent with torch baseline benchmarks.

Priya2698 replied:
Dropout: 0.2
![Screenshot 2024-04-04 at 11 53 57 AM](https://github.com/NVIDIA/Fuser/assets/52657555/0adf9ff0-811b-4512-a6a9-1c958669ca50)

Dropout: 0.0
![Screenshot 2024-04-04 at 11 53 49 AM](https://github.com/NVIDIA/Fuser/assets/52657555/ee1ef950-c084-46b4-a93b-f237021774c4)

The difference for eager is significant.

liqiangxl replied:
> What do you think about having another fusion definition with 0.0 dropout in validation and 0.2 otherwise. That would ensure we always benchmark with dropout (0.2) and it is consistent with torch baseline benchmarks.

sounds good to me.

liqiangxl replied:
> Dropout: 0.2 ![Screenshot 2024-04-04 at 11 53 57 AM](https://private-user-images.githubusercontent.com/52657555/319736946-0adf9ff0-811b-4512-a6a9-1c958669ca50.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIyNTkyOTksIm5iZiI6MTcxMjI1ODk5OSwicGF0aCI6Ii81MjY1NzU1NS8zMTk3MzY5NDYtMGFkZjlmZjAtODExYi00NTEyLWE2YTktMWM5NTg2NjljYTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA0VDE5Mjk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5YmRmMjUzNzZhYTc1ODhkMjY0YzUxYTkyZTNjMmZiZGJkNmEyZjk4Mzk0ZmFmOTYxNzc0ZmY1ZjBhZTQ2MzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.g9kgsWCSZF_WksLoaZMgAuHcSCWYQmrdlfAW4yoGFfg)
> 
> Dropout: 0.0 ![Screenshot 2024-04-04 at 11 53 49 AM](https://private-user-images.githubusercontent.com/52657555/319736994-ee1ef950-c084-46b4-a93b-f237021774c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIyNTkyOTksIm5iZiI6MTcxMjI1ODk5OSwicGF0aCI6Ii81MjY1NzU1NS8zMTk3MzY5OTQtZWUxZWY5NTAtYzA4NC00NmI0LWE5M2ItZjIzNzAyMTc3NGM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA0VDE5Mjk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMzM5OGMyMTllZjg3MWFjZjBmMmU0ZjI1MDM4YjczYzRkYTMyYzNjZTlhNWQ2MGNhMzcyOGU0ZGE2ZDViZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OWxJoj0beWfZvLpahf70eOveRZzpxRuo62FyprTdBXU)
> 
> The difference for eager is significant.

This seems difficult to explain. can you create an issue to investigate further when we have time?

jacobhinkle replied:
Eager mode is probably skipping the dropout call altogether when `dropout=0` but i would guess torch.compile is either not skipping dropout, or it's just io-bound so skipping dropout makes no difference.

liqiangxl replied:
that's very smart!

jacobhinkle commented on benchmarks/python/test_huggingface_attn_bwd.py:
```diff
@@ -55,9 +55,15 @@ def huggingface_attn_bwd_fusion(
     fd.add_output(T35)
 
 
+def huggingface_attn_bwd_iobytes(size: tuple, dtype: torch.dtype):
```
tiny nit: when annotating collections, specify the element types. For tuple, annotate all of them like `size: tuple[int, int, int, int]`.

jacobhinkle replied:
As @liqiangxl already mentioned, if this just implements the default behavior i.e. `sum(i.nbytes for i in inputs) + sum(o.nbytes for o in outputs)` then we should remove this helper.

Priya2698 replied:
This is for uniformity of IObytes between nvFuser and torch when the inputs/outputs differ between the two. See the discussion at: https://github.com/NVIDIA/Fuser/pull/1725. I'll add a reference to this PR in a comment.

Priya2698 replied:
> tiny nit: when annotating collections, specify the element types. For tuple, annotate all of them like `size: tuple[int, int, int, int]`.

I think for size, the element types are understood. So I prefer this to reduce verbosity. Not a strong preference, though. I may have a bias being more familiar with the benchmarks.

jacobhinkle replied:
Does the IOBytes issue affect this computation though? It seems like the helper method is just replicating default behavior. If that's the case, then maybe putting argument `iobytes=None,` would be the clearest. As you wish on the tuple annotation.

Priya2698 replied:
The nvFuser input/outputs are grad_out, attention_mask, dropout_mask and grad_input.
The eager input/outputs are grad_out and output. So there will be a difference in the computed IObytes.

jacobhinkle replied:
Oh ok, so it does affect this case due to recomputation. Could you place a comment in the code describing that? Thanks.

jacobhinkle commented on benchmarks/python/test_huggingface_attn_bwd.py:
```diff
@@ -95,3 +101,34 @@ def test_huggingface_attn_bwd_benchmark(
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_huggingface_attn_bwd_baseline_benchmark(
+    benchmark,
+    size: tuple,
```
Nit: `size: tuple[int, int, int, int]`

jacobhinkle commented on benchmarks/python/test_huggingface_attn_fwd.py:
```diff
@@ -71,9 +71,26 @@ def huggingface_attn_fwd_fusion(
 
+def huggingface_attn_fwd_fn(
+    inputs: list,
+):  # [attention_mask, in_tensor, size, dropout_p]
```
Comment should be a docstring if providing more details, or omitted since it's clear from the following line.

jacobhinkle commented on benchmarks/python/test_huggingface_attn_fwd.py:
```diff
@@ -71,9 +71,26 @@ def huggingface_attn_fwd_fusion(
     fd.add_output(T47)
 
 
+def huggingface_attn_fwd_fn(
```
Might I suggest
```suggestion
def huggingface_attn_fwd(
```
And related changes further down in this file.

Priya2698 commented:
Updated the PR description.
We use a non-zero dropout in benchmarking now. For fwd calls, I set it to 0 for validation.

Discovered a bug when using non-zero dropout in nanogpt bwd pass that is now fixed.
